### DISCUSSION TITLE
Same route for different controllers depending on the logged user

### DISCUSSION THREAD

**Discussion by realfabiopio** [Laravel-Backpack/community-forum#964] [Q&A (Help)]
I need to implement a solution that allows the same route for different controllers depending on the logged user.
Some users have customizations in their controllers, those who do not must use the route's default controller.
In LoginController I store the conditional in the session.

Example:

Controllers:
 - App\Http\Controllers\abcController.php
 - App\Http\Controllers\C0007\abcController.php - extends App\Http\Controllers\abcController
 - App\Http\Controllers\C0012\abcController.php - extends App\Http\Controllers\abcController
 and so on...

route:
  Route::crud('abc', 'App\Http\Controllers\abcController');

I tried to use a middleware but any attempt to change the controller caused the object to need to be instantiated and thus lost all references (e.g. in setupListOperation() $this->crud was null)

Is there a way to solve this?

**Comment by pxpm** [comment]
@realfabiopio hey there. 

I think you are approaching the problem from the wrong perspective. I think it would be easier and more correct if the route always resolve to a "BaseController", in that "BaseController" you can check the authenticated user and call the appropriate functions. 
If you still want to keep a similar "structure" you can do something like: 
```php
// App\Actions\C001Action.php

public function __invoke() {
// if you need crud here, you can resolve it from the container
$crud = app('crud');
}

// App\Actions\C0012Action.php

public function __invoke() {
// ...
}

// and your base controller
public function endpoint()
{
return match($user->role) {
'admin' => new \App\Actions\C001Action(),
'user' => new \App\Actions\C0012Action()
}
```

If you are talking only about permissions, you should do that check in your controller, using a Gate or whatever you feel better using. 
Routes are setup BEFORE there is any authenticated user, I never seen what you want to achieve be applied in any project, and it doesn't feel right to do that. 

Like I proposed is more common to see, but usually there is no need to do that in CRUD controllers as actions are more suited for single action controllers. 

Cheers

**Reply by realfabiopio** [reply]
The descendant controllers are normal backpack controllers, they have some different methods than the base ones, different columns and different forms according to the customer's needs... 
I don't know what the best solution is to solve this. It was suggested to me to change the request and its action in a middleware and I even used your idea of ​​getting the crud obj from app('crud') but the app('crud') ​​always returns me the crud of the base controller, that is, even if I change the action to a descendant controller, the crud is from the controller used in the route, the base one.
Another (crazy) idea would be to put all the different methods and countless ifs in a single controller and create a crazy spider web.

What I really needed is to organize the customizations of my customers in diferent files as I mentioned before (a base controller and descendants according to the necessary customization, each with their respective file)

What would be the best way to solve this? Rephrasing the question another way:
How can I make the same route point to descendant controllers (in different files) according to the logged in user?

**Comment by jcastroa87** [comment]
Hi @realfabiopio 

I didn't test this, but in theory.

Create a Middleware

```
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;

class RedirectBasedOnUserRole
{
    public function handle(Request $request, Closure $next)
    {
        // Check if the user is logged in
        if (Auth::check()) {
            $user = Auth::user();

            // Check user's role and redirect accordingly
            if ($user->isAdmin()) {
                return redirect()->route('admin.dashboard');
            } elseif ($user->isCustomer()) {
                return redirect()->route('customer.dashboard');
            }
        }

        // For other cases, proceed with the request
        return $next($request);
    }
}
```

on routes
```
// Step 2: Group Routes

Route::middleware(['auth', 'role.redirect'])->group(function () {
    Route::get('/dashboard', 'DashboardController@index')->name('admin.dashboard')->middleware('admin');
    Route::get('/dashboard', 'CustomerDashboardController@index')->name('customer.dashboard')->middleware('customer');
});
```

Check is that works for you.

Cheers.

**Comment by jcastroa87** [comment]
Due to not activity i will close this issue, but please feel free to re-open or create a new one if needed.

Cheers.
